<a href="https://colab.research.google.com/github/channel1umut-lang/Machine-Learning-Applications/blob/main/UMUT_YAVAS_ML_FINAL_M505A.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# TITLE : MACHINE LEARNING PIPELINE FOR WATER QUALITY PREDICTION

# 1. PROBLEM STATEMENT
### Business Problem
To access clean drinking water is a fundamental of human needs. Contaminated water can lead to serious diseases and health complications. Water quality is determined by various physical, chemical, and microbiological parameters. However, traditional laboratory testing methods are time-consuming, expensive, and not easily accessible in many regions.

Our company provides consulting services to municipalities and water treatment facilities. Currently, our clients face challenges in quickly and cost-effectively assessing whether water is potable (safe for human consumption). They need a reliable, fast, and scalable solution to evaluate water quality.

### Why Is This Important?
- **Public Health:** Ensuring water safety prevents waterborne diseases and protects community health
- **Regulatory Compliance:** Municipalities must meet strict water quality standards
- **Operational Efficiency:** Water treatment plants need real-time monitoring to optimize their processes
- **Cost Reduction:** Traditional testing requires expensive laboratory equipment and trained personnel

### How Will Solving This Problem Benefit the Company?
1. **Competitive Advantage:** Offering AI-powered water quality assessment will differentiate us from competitors who rely only on traditional methods
2. **New Revenue Stream:** We can provide this as a subscription-based service to municipalities and water facilities
3. **Scalability:** Once developed, the solution can be deployed to multiple clients with minimal additional cost
4. **Data-Driven Insights:** Accumulated data can help identify trends and predict potential water quality issues before they become critical


### Project Objective
To develop a machine learning pipeline that accurately predicts water potability, enabling rapid, cost-effective water quality assessment.

In [ ]:
import pandas as pd
import numpy as np
import sklearn
from IPython.display import display
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import GridSearchCV , RandomizedSearchCV
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier

In [ ]:
#https://www.kaggle.com/datasets/adityakadiwal/water-potability/code

# 2. DATA COLLECTION STRATEGY

## Dataset Source
The dataset used in this project is sourced from Kaggle: "Water Quality" dataset by Aditya Kadiwal. It contains 3,276 water samples with 9 water quality parameters and a target variable indicating water potability.

## Real-World Data Collection Methods
- IoT sensors installed at water sources and treatment plants measure parameters in real-time
- Laboratory tests analyze water samples for chemical composition to ensure accuracy
- Historical records from municipalities provide valuable baseline data and long-term trends
- Partnerships with environmental agencies allow access to broader regional datasets



In [ ]:
df = pd.read_csv('https://www.kaggle.com/api/v1/datasets/download/adityakadiwal/water-potability/water_potability.csv')

# 3. DATA EXPLORATION
# Dataset Overview
The dataset contains **3,276 water samples** with **10 features** (9 input features and 1 target variable). Each sample represents water quality measurements from different sources.


## Water Quality Parameters
###Features:
- **pH Level:** Measures acidity on a scale of 0-14, with 6.5-8.5 considered safe for drinking
- **Hardness:** Concentration of calcium and magnesium salts measured in mg/L
- **Total Dissolved Solids (TDS):** Concentration of dissolved substances in water measured in ppm
- **Chloramines:** Disinfectants used to control bacteria, formed by combining chlorine and ammonia
- **Sulfate:** Naturally occurring mineral that can affect taste at high concentrations
- **Conductivity:** Water's ability to conduct electricity, indicating ion concentration
- **Organic Carbon:** Amount of organic matter that can serve as food for microorganisms
- **Trihalomethanes:** Compounds formed when chlorine reacts with organic matter
- **Turbidity:** Cloudiness caused by suspended particles measured in NTU

### Target Variable
- **Potability:** Binary indicator where 1 = safe for drinking, 0 = not safe for drinking

## Data Dictionary:
 **pH:** Acidity level (0-14 scale)

 **Hardness:** Calcium and magnesium concentration (mg/L)

 **Solids:** Total dissolved solids (ppm)

 **Chloramines:** Disinfectant concentration (ppm)

 **Sulfate:** Sulfate mineral concentration (mg/L)

 **Conductivity:** Electrical conductivity (μS/cm)

 **Organic_carbon:** Organic matter content (mg/L)

 **Trihalomethanes:** THM compounds concentration (μg/L)

 **Turbidity:** Water clarity measure (NTU)

 **Potability:** Target variable (0 or 1)


In [ ]:
display(df)

,ph,Hardness,Solids,Chloramines,Sulfate,Conductivity,Organic_carbon,Trihalomethanes,Turbidity,Potability
0,NaN,204.890455,20791.318981,7.300212,368.516441,564.308654,10.379783,86.990970,2.963135,0
1,3.716080,129.422921,18630.057858,6.635246,NaN,592.885359,15.180013,56.329076,4.500656,0
2,8.099124,224.236259,19909.541732,9.275884,NaN,418.606213,16.868637,66.420093,3.055934,0
3,8.316766,214.373394,22018.417441,8.059332,356.886136,363.266516,18.436524,100.341674,4.628771,0
4,9.092223,181.101509,17978.986339,6.546600,310.135738,398.410813,11.558279,31.997993,4.075075,0
...,...,...,...,...,...,...,...,...,...,...
3271,4.668102,193.681735,47580.991603,7.166639,359.948574,526.424171,13.894419,66.687695,4.435821,1
3272,7.808856,193.553212,17329.802160,8.061362,NaN,392.449580,19.903225,NaN,2.798243,1
3273,9.419510,175.762646,33155.578218,7.350233,NaN,432.044783,11.039070,69.845400,3.298875,1
3274,5.126763,230.603758,11983.869376,6.303357,NaN,402.883113,11.168946,77.488213,4.708658,1


In [ ]:
df.shape

(3276, 10)

In [ ]:
df.columns

Index(['ph', 'Hardness', 'Solids', 'Chloramines', 'Sulfate', 'Conductivity',
       'Organic_carbon', 'Trihalomethanes', 'Turbidity', 'Potability'],
      dtype='object')

#4. TRAIN-TEST SPLIT

The dataset is split into training and testing sets before any preprocessing to prevent data leakage. This ensures that information from the test set does not influence the training process.


In [ ]:
df_train, df_test = sklearn.model_selection.train_test_split(df, random_state=15)

print("df size:", df.shape)
print("df_train size:", df_train.shape)
print("df_test size:", df_test.shape)

df size: (3276, 10)
df_train size: (2457, 10)
df_test size: (819, 10)


# 5. Missing Value Analysis and Data Type Inspection
###Data Type Inspection
Before handling missing values, we examine the data types of each column to ensure they are appropriate for analysis. All features are numeric (float64) and the target variable Potability is integer (int64), which is suitable for classification tasks. No data type conversion is needed.

###Missing Value Analysis
After splitting the data, we check for missing values in both training and test sets. The analysis reveals that three columns contain null values:

The ph column has 368 missing values in training set, 123 missing values in test set.

The Sulfate column has 592 missing values in training set, 189 missing values in test set

Trihalomethanes column has 127 missing values in training set, 35 missing values in test set

The rest of the columns do not have missing values.

In [ ]:
display(df_train.info()), display(df_test.info())

<class 'pandas.core.frame.DataFrame'>
Index: 2457 entries, 255 to 2693
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   ph               2089 non-null   float64
 1   Hardness         2457 non-null   float64
 2   Solids           2457 non-null   float64
 3   Chloramines      2457 non-null   float64
 4   Sulfate          1865 non-null   float64
 5   Conductivity     2457 non-null   float64
 6   Organic_carbon   2457 non-null   float64
 7   Trihalomethanes  2330 non-null   float64
 8   Turbidity        2457 non-null   float64
 9   Potability       2457 non-null   int64  
dtypes: float64(9), int64(1)
memory usage: 211.1 KB


None

<class 'pandas.core.frame.DataFrame'>
Index: 819 entries, 3126 to 1716
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   ph               696 non-null    float64
 1   Hardness         819 non-null    float64
 2   Solids           819 non-null    float64
 3   Chloramines      819 non-null    float64
 4   Sulfate          630 non-null    float64
 5   Conductivity     819 non-null    float64
 6   Organic_carbon   819 non-null    float64
 7   Trihalomethanes  784 non-null    float64
 8   Turbidity        819 non-null    float64
 9   Potability       819 non-null    int64  
dtypes: float64(9), int64(1)
memory usage: 70.4 KB


None

(None, None)

In [ ]:
display(df_train.isnull().sum()), display(df_test.isnull().sum())


,0
ph,368
Hardness,0
Solids,0
Chloramines,0
Sulfate,592
Conductivity,0
Organic_carbon,0
Trihalomethanes,127
Turbidity,0
Potability,0


,0
ph,123
Hardness,0
Solids,0
Chloramines,0
Sulfate,189
Conductivity,0
Organic_carbon,0
Trihalomethanes,35
Turbidity,0
Potability,0


(None, None)

#6. Handling Missing Values
To address these missing values without causing data leakage, we calculate the median for each column using only the training set. The median is chosen over the mean because it is less sensitive to outliers. These median values are then used to fill missing entries in both training and test sets. This approach prevents data leakage issues and fills nan values with logical numbers in both train and test set.

### Verification
After filling, we verify that no missing values remain in either dataset. Both training and test sets now have complete data across all columns, ready for further analysis and model development.

In [ ]:
median_ph = df_train['ph'].median()
median_sulfate = df_train['Sulfate'].median()
median_tri = df_train['Trihalomethanes'].median()

df_train['ph'] = df_train['ph'].fillna(median_ph)
df_train['Sulfate'] = df_train['Sulfate'].fillna(median_sulfate)
df_train['Trihalomethanes'] = df_train['Trihalomethanes'].fillna(median_tri)

df_test['ph'] = df_test['ph'].fillna(median_ph)
df_test['Sulfate'] = df_test['Sulfate'].fillna(median_sulfate)
df_test['Trihalomethanes'] = df_test['Trihalomethanes'].fillna(median_tri)

In [ ]:
display(df_train.isnull().sum()), display(df_test.isnull().sum())

,0
ph,0
Hardness,0
Solids,0
Chloramines,0
Sulfate,0
Conductivity,0
Organic_carbon,0
Trihalomethanes,0
Turbidity,0
Potability,0


,0
ph,0
Hardness,0
Solids,0
Chloramines,0
Sulfate,0
Conductivity,0
Organic_carbon,0
Trihalomethanes,0
Turbidity,0
Potability,0


(None, None)

In [ ]:
display(df_train.describe()) , display(df_test.describe())

,ph,Hardness,Solids,Chloramines,Sulfate,Conductivity,Organic_carbon,Trihalomethanes,Turbidity,Potability
count,2457.000000,2457.000000,2457.000000,2457.000000,2457.000000,2457.000000,2457.000000,2457.000000,2457.000000,2457.000000
mean,7.048429,196.742488,21995.932556,7.135767,333.496026,426.839361,14.340332,66.687771,3.972576,0.378510
std,1.470732,33.212106,8718.861651,1.559338,35.831424,80.967608,3.297007,15.690854,0.781395,0.485114
min,0.000000,47.432000,320.942611,0.530351,129.000000,181.483754,2.200000,0.738000,1.450000,0.000000
25%,6.266800,177.233643,15629.762335,6.190963,316.675162,367.389811,12.083896,57.247082,3.450746,0.000000
50%,7.007975,197.331291,20886.786563,7.143907,332.766156,421.464253,14.233952,67.052599,3.968177,0.000000
75%,7.831118,216.782881,27267.981861,8.114731,350.363703,484.172953,16.589072,76.831226,4.497731,1.000000
max,14.000000,323.124000,56867.859236,13.127000,462.474215,753.342620,28.300000,124.000000,6.739000,1.000000


,ph,Hardness,Solids,Chloramines,Sulfate,Conductivity,Organic_carbon,Trihalomethanes,Turbidity,Potability
count,819.000000,819.000000,819.000000,819.000000,819.000000,819.000000,819.000000,819.000000,819.000000,819.000000
mean,7.134234,195.250519,22068.572436,7.081806,333.652251,424.302361,14.118885,65.651679,3.949416,0.424908
std,1.467562,31.855601,8921.202276,1.652601,37.092525,80.411149,3.337921,15.991655,0.777555,0.494631
min,1.757037,73.492234,1351.906979,0.352000,180.206746,210.319182,4.861631,8.577013,1.492207,0.000000
25%,6.338633,175.586910,15762.263670,5.997201,318.429686,362.637927,11.976289,55.314102,3.413018,0.000000
50%,7.007975,195.018888,21038.442613,7.066565,332.766156,422.817556,14.184036,66.291650,3.917587,0.000000
75%,7.949652,215.926973,27462.947300,8.114066,350.389532,477.129673,16.374668,76.449490,4.515554,1.000000
max,11.898078,287.975540,61227.196008,13.043806,481.030642,708.226364,23.917601,120.030077,6.494749,1.000000


(None, None)

In [ ]:
display(df_train.Potability.value_counts()), display(df_test.Potability.value_counts())

,count
Potability,
0,1527
1,930


,count
Potability,
0,471
1,348


(None, None)

#7. Feature Selection and Scaling

### Separating Features and Target Variable
Before model training, I separated the features from the target variable. The target variable "Potability" is removed from the feature sets and stored separately for both training and testing data.

- **X_train:** Training features (2457 samples, 9 parameters)
- **X_test:** Testing features (819 samples, 9 parameters)
- **y_train:** Training target values (2457 samples)
- **y_test:** Testing target values (819 samples)
###Feature Scaling
 I applied StandardScaler to standardize the features The scaler is fitted only on the training data to prevent data leakage, then used to transform both
training and test sets.


In [ ]:
X_train = df_train.drop("Potability", axis=1)
X_test = df_test.drop("Potability", axis=1)
y_train = df_train["Potability"]
y_test = df_test["Potability"]

print("x_train size:", X_train.shape)
print("x_test size:", X_test.shape)
print("y_train size:", y_train.shape)
print("y_test size:", y_test.shape)

x_train size: (2457, 9)
x_test size: (819, 9)
y_train size: (2457,)
y_test size: (819,)


In [ ]:
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

#8. MODEL TRAING
For model development, it is selected four different classification algorithms to predict water potability. Each model was tuned using cross-validation (5-fold) to find the best hyperparameters. The training was performed on the scaled feature set (X_train_scaled) with the target variable (y_train).

#Preferred models:
**1. K-Nearest Neighbors – Hyperparameter Tuning with GridSearchCV**

**2. Decision Tree – Hyperparameter Tuning with GridSearchCV**

**3. Logistic Regression – Hyperparameter Tuning with GridSearchCV**

**4. Random Forest – Hyperparameter Tuning with RandomizedSearchCV**

##8.1. K-NEAREST NEIGHBORS CLASSIFIER
 I performed hyperparameter tuning using GridSearchCV with 5-fold cross-validation to find the optimal configuration.
Best Parameters: {'n_neighbors': 9, 'weights': 'distance'}
Best Cross-Validation Score: 0.6455

The optimal configuration uses 9 neighbors with distance-based weighting, meaning closer neighbors have a greater influence on the prediction than those farther away. This approach achieved a cross-validation accuracy of 64.55%.

In [ ]:
knn_params = {
    'n_neighbors': [3, 5, 7, 9],
    'weights': ['uniform', 'distance']
}

In [ ]:
knn_grid = GridSearchCV(estimator = KNeighborsClassifier(), param_grid=knn_params, cv=5,scoring="accuracy")

In [ ]:
knn_grid.fit(X_train_scaled,y_train)

GridSearchCV(cv=5, estimator=KNeighborsClassifier(),
             param_grid={'n_neighbors': [3, 5, 7, 9],
                         'weights': ['uniform', 'distance']},
             scoring='accuracy')

In [ ]:
knn_grid.best_params_

{'n_neighbors': 9, 'weights': 'distance'}

In [ ]:
knn_grid.best_score_

np.float64(0.6455160366267614)

##8.2. DECISION TREES
I performed hyperparameter tuning using GridSearchCV with 5-fold cross-validation to find the optimal configuration.
Best Parameters: {'criterion': 'gini', 'max_depth': 5, 'max_features': None, 'splitter': 'best'}

Best Cross-Validation Score: 0.6435

The optimal configuration uses a maximum depth of 5 with gini impurity as the splitting criterion, considering all features at each split. This approach achieved a cross-validation accuracy of 64.35%.

In [ ]:
dt_param = {
    "criterion" : ["gini", "entropy", "log_loss"],
    "splitter" : ["best", "random"],
    "max_depth" : [1,2,3,4,5,15,None],
    "max_features" : ["sqrt", "log2", None]
}

In [ ]:
dt_grid = GridSearchCV(estimator = DecisionTreeClassifier(random_state=42), param_grid=dt_param, cv=5,scoring="accuracy")
dt_grid.fit(X_train_scaled, y_train)

GridSearchCV(cv=5, estimator=DecisionTreeClassifier(random_state=42),
             param_grid={'criterion': ['gini', 'entropy', 'log_loss'],
                         'max_depth': [1, 2, 3, 4, 5, 15, None],
                         'max_features': ['sqrt', 'log2', None],
                         'splitter': ['best', 'random']},
             scoring='accuracy')

In [ ]:
dt_grid.best_params_

{'criterion': 'gini', 'max_depth': 5, 'max_features': None, 'splitter': 'best'}

In [ ]:
dt_grid.best_score_

np.float64(0.643473581375325)

##8.3. LOGISTIC REGRESSION-->LOWEST CV SCORE!
I performed hyperparameter tuning using GridSearchCV with 5-fold cross-validation to find the optimal configuration.
Best Parameters: {'C': 0.1, 'penalty': 'l2', 'solver': 'liblinear'}

Best Cross-Validation Score: 0.6247

The optimal configuration uses a regularization strength of 0.1 with L2 penalty, optimized using the liblinear solver. This approach achieved a cross-validation accuracy of 62.47%.

In [ ]:
logistic_params = {
    'C': [0.01, 0.1, 1, 10, 100],
    'penalty': ['l1', 'l2'],
    'solver': ['liblinear']
}

In [ ]:
logistic_grid = GridSearchCV(estimator = LogisticRegression(), param_grid=logistic_params, cv=5,scoring="accuracy")
logistic_grid.fit(X_train_scaled, y_train)

GridSearchCV(cv=5, estimator=LogisticRegression(),
             param_grid={'C': [0.01, 0.1, 1, 10, 100], 'penalty': ['l1', 'l2'],
                         'solver': ['liblinear']},
             scoring='accuracy')

In [ ]:
logistic_grid.best_params_

{'C': 0.1, 'penalty': 'l2', 'solver': 'liblinear'}

In [ ]:
logistic_grid.best_score_

np.float64(0.6247462454257944)

##8.4. RANDOM FOREST --> HIGHEST CV SCORE!

 Given the larger hyperparameter space, I used RandomizedSearchCV with 5-fold cross-validation to efficiently find the optimal configuration.

Best Parameters: {'n_estimators': 500, 'min_samples_split': 3, 'max_features': None, 'max_depth': 10}

Best Cross-Validation Score: 0.6724

The optimal configuration uses 500 trees with a maximum depth of 10, considering all features at each split, and requiring a minimum of 3 samples to split an internal node. This approach achieved the highest cross-validation accuracy among all models at 67.24%.

In [ ]:
forest_Classsifier = RandomForestClassifier(random_state=15)

In [ ]:
rf_params = {
    "max_depth": [5, 10, 15, None],
    "max_features": [5, 8, 'log2', None],
    "min_samples_split": [3, 10, 15],
    "n_estimators": [100, 200, 500]
}

In [ ]:
rscv = RandomizedSearchCV(estimator=forest_Classsifier, param_distributions=rf_params,n_iter=10,cv=5, random_state=42, n_jobs=-1)
rscv.fit(X_train_scaled, y_train)

RandomizedSearchCV(cv=5, estimator=RandomForestClassifier(random_state=15),
                   n_jobs=-1,
                   param_distributions={'max_depth': [5, 10, 15, None],
                                        'max_features': [5, 8, 'log2', None],
                                        'min_samples_split': [3, 10, 15],
                                        'n_estimators': [100, 200, 500]},
                   random_state=42)

In [ ]:
rscv.best_params_

{'n_estimators': 500,
 'min_samples_split': 3,
 'max_features': None,
 'max_depth': 10}

In [ ]:
rscv.best_score_

np.float64(0.6723684864139885)

#9. MODEL EVALUATION – RANDOM FOREST (BEST MODEL)

After comparing all four models, Random Forest achieved the highest cross-validation score (0.6724). We selected it as our final model and evaluated its performance on the unseen test set.

The model achieved a test accuracy of 0.6703, which is very close to the cross-validation score, indicating good generalization without overfitting.

The confusion matrix shows that the model correctly identified 445 non-potable samples and 104 potable samples, while misclassifying 26 non-potable and 244 potable samples.

The classification report reveals that the model performs very well on class 0 (non-potable) with 94% recall, but struggles with class 1 (potable) achieving only 30% recall. This suggests the model is better at detecting unsafe water than confirming safe water.

In [ ]:
best_rf = rscv.best_estimator_

In [ ]:
y_pred_RF = best_rf.predict(X_test_scaled)

In [ ]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
print(accuracy_score(y_test, y_pred_RF))
print(confusion_matrix(y_test, y_pred_RF))
print(classification_report(y_test, y_pred_RF))

0.6703296703296703
[[445  26]
 [244 104]]
              precision    recall  f1-score   support

           0       0.65      0.94      0.77       471
           1       0.80      0.30      0.44       348

    accuracy                           0.67       819
   macro avg       0.72      0.62      0.60       819
weighted avg       0.71      0.67      0.63       819



# 10. CONCLUSION AND DISCUSSION

## Summary of Findings
In this project, four different classification algorithms were developed to predict water potability. Random Forest was selected as the best-performing model, achieving a cross-validation score of 0.6724 and a test accuracy of 0.6703.

However, a closer look at the model's performance reveals a significant imbalance:
- **Non-potable water (class 0):** 94% correctly identified (445 out of 471)
- **Potable water (class 1):** Only 30% correctly identified (104 out of 348)

These results indicate that the model cannot be used alone in real-world applications. A 70% error rate in identifying safe drinking water poses an unacceptable public health risk.

## Business Implications
The model may only be useful in the following contexts:
- As a **screening tool** to quickly flag clearly unsafe water, reducing the number of samples requiring full laboratory analysis.
- In combination with existing laboratory tests to improve operational efficiency.
- As a **prototype** requiring further development before deployment.

## Final Remarks
While the current model does not meet the reliability standards for standalone deployment, it successfully demonstrates the feasibility of using machine learning for water quality screening. With further refinement and domain-specific tuning, this solution has the potential to deliver significant value to our clients while contributing to public health and safety.

In [ ]:
#only this part of code was created by AI tool in order to count the number of the character
from google.colab import drive
drive.mount('/content/drive')

import json

notebook_path = '/content/drive/MyDrive/Colab Notebooks/UMUT_YAVAS_ML_FINAL_M505A.ipynb'

with open(notebook_path, 'r', encoding='utf-8') as f:
    notebook = json.load(f)

total_chars = sum(len(''.join(cell['source'])) for cell in notebook['cells'])
print(f"Total characters: {total_chars}")

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
Total characters: 15802


In [ ]:
!jupyter nbconvert --to html UMUT_YAVAS_ML_FINAL_M505A.ipynb

[NbConvertApp] Converting notebook UMUT_YAVAS_ML_FINAL_M505A.ipynb to html
[NbConvertApp] Writing 437286 bytes to UMUT_YAVAS_ML_FINAL_M505A.html
